### ISSUE TITLE
[PrimTorch] Invalid partition found when running TIMM hrnet_w18

### ISSUE THREAD

**Issue by rdspring1** [pytorch/pytorch#86159]
### 🐛 Describe the bug

Running into issue during tracing:

```Python
Traceback (most recent call last):
  File "/home/rds4/workspace/nvprims-torchdynamo/benchmarks/common.py", line 1171, in warmup
    fn(model, example_inputs)
  File "/home/rds4/anaconda3/lib/python3.8/site-packages/torchdynamo-1.13.0.dev0-py3.8-linux-x86_64.egg/torchdynamo/eval_frame.py", line 165, in _fn
    return fn(*args, **kwargs)
  File "benchmarks/timm_models.py", line 317, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "benchmarks/timm_models.py", line 318, in <graph break in forward_and_backward_pass>
    cloned_inputs = clone_inputs(inputs)
  File "benchmarks/timm_models.py", line 319, in <graph break in forward_and_backward_pass>
    mod.zero_grad(True)
  File "benchmarks/timm_models.py", line 321, in <graph break in forward_and_backward_pass>
    pred = mod(*cloned_inputs)
  File "/home/rds4/workspace/my_pytorch/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/rds4/anaconda3/lib/python3.8/site-packages/timm-0.6.5-py3.8.egg/timm/models/hrnet.py", line 748, in forward
    def forward(self, x):
  File "/home/rds4/anaconda3/lib/python3.8/site-packages/torchdynamo-1.13.0.dev0-py3.8-linux-x86_64.egg/torchdynamo/eval_frame.py", line 165, in _fn
    return fn(*args, **kwargs)
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 797, in forward
    return compiled_f(
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 788, in new_func
    return compiled_fn(args)
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 229, in g
    return f(*args)
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 434, in compiled_function
    return CompiledFunction.apply(*remove_dupe_args(args))
  File "/home/rds4/anaconda3/lib/python3.8/site-packages/torchdynamo-1.13.0.dev0-py3.8-linux-x86_64.egg/torchdynamo/eval_frame.py", line 165, in _fn
    return fn(*args, **kwargs)
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 408, in forward
    fw_outs = call_func_with_args(
  File "/home/rds4/workspace/my_pytorch/functorch/_src/aot_autograd.py", line 260, in call_func_with_args
    out = normalize_as_list(f(*args))
  File "/home/rds4/workspace/my_pytorch/torch/_prims/executor.py", line 25, in execute
    return nvfuser_execute_partitioned(
  File "/home/rds4/workspace/my_pytorch/torch/_prims/nvfuser_executor.py", line 343, in nvfuser_execute_partitioned
    gm, is_partitioned = maybe_partition_graph(
  File "/home/rds4/workspace/my_pytorch/torch/_prims/nvfuser_executor.py", line 310, in maybe_partition_graph
    partitioned_graph = partitioner.fuse_partitions(partitions)
  File "/home/rds4/workspace/my_pytorch/torch/fx/passes/infra/partitioner.py", line 222, in fuse_partitions
    return fuse_by_partitions(self.graph_module, [list(partition.nodes) for partition in partitions])
  File "/home/rds4/workspace/my_pytorch/torch/fx/passes/utils/fuser_utils.py", line 204, in fuse_by_partitions
    sub_gm, orig_inputs, orig_outputs = fuse_as_graphmodule(gm, sorted_nodes, submodule_name)
  File "/home/rds4/workspace/my_pytorch/torch/fx/passes/utils/fuser_utils.py", line 106, in fuse_as_graphmodule
    assert validate_partition(nodes), "Invalid partition, found dependency cycles"
AssertionError: Invalid partition, found dependency cycles
```

Repro Command:
`python -u benchmarks/timm_models.py --training -d cuda --fast --backend nvprims_nvfuser --skip-accuracy-check --performance  --generate-aot-autograd-stats --dump-raw-metrics --amp -k hrnet_w18`

### Versions

Pytorch
commit 6d06be8

TorchDynamo
commit a043fcc

**Comment by jjsjann123** [comment]
Mark this issue since I'm looking at graph partitioner.

**Comment by jjsjann123** [comment]
Looks like there's indeed a circular dependency, something is not right with the partitioner.
Graph is really big and I haven't figured out exactly what's wrong with current partitioning logic yet.

**Comment by jjsjann123** [comment]
> Looks like there's indeed a circular dependency, something is not right with the partitioner. Graph is really big (~6k nodes) crying_cat_face and I haven't figured out exactly what's wrong with current partitioning logic yet.

Since I can't wrap my head around what's currently wrong with the implementation, one quick way to debug this is to go through cyclic check for each merge partition. At least that should point me where it happened and maybe I'll come up with a counter example.

Alternatively, we can just update the partitioning logic to use cyclic check directly at least we know it's always safe to do that. :face_exhaling: 

cc'ing @SherlockNoMad

**Comment by jjsjann123** [comment]
Finally worked out a repro. The bug comes from our dependency map, we should be updating dependency map per partition as we propose new partitions.
Dependency map is not "static" as we fuse nodes. [https://github.com/pytorch/pytorch/blob/91b1bae1df1e72e17d2ab296845c214bc39422a0/torch/fx/passes/infra/partitioner.py#L49-L59]
When independent nodes are fused, their dependency map should be aggregated to properly reflect the new topo relationship after the fusion. Otherwise, we are ignoring the aggregated dependencies of partitions and would create cyclic graph.

Updating the aggregated dependency is going to be expensive. Especially we are also kicking nodes out of fusion. I'll see if I can patch it in a reasonable way.

Also noticed some mismatch between operator support and partitioner. Looks like we had two patches in place to merge `getitem`:
https://github.com/pytorch/pytorch/blob/91b1bae1df1e72e17d2ab296845c214bc39422a0/torch/fx/passes/infra/partitioner.py#L177-L195
https://github.com/pytorch/pytorch/blob/master/torch/_prims/nvfuser_executor.py#L243

These two added redundant work where we fuse `getitem` as input to partition and later delete them from the partition. :shrug:

**Comment by jjsjann123** [comment]
Started a draft with a repro. Will update it once I got the fix in.

**Comment by jjsjann123** [comment]
Properly patched and fixed in #86511 

Already pinged Sherlock to look at the PR.
